### PR TITLE
Horizontal rule support

### PR DESCRIPTION
**Pull request by daviscabral**
/cc @marconilanna

### CODE REVIEW

File: REPLesent.scala [hunk 1]
```diff
@@ -92,6 +92,29 @@ case class REPLesent(
       def apply(line: Line, margin: Int): String
     }
 
+    protected sealed trait Repeat extends Alignment {
+      protected def fill(line: Line, width: Int, left: Int, right: Int): String = {
```
Comment (marconilanna) [inline REPLesent.scala]: You could get rid of this trait, incorporate this method into `HorizontalRuler`, and have `FullScreenHorizontalRuler` simply call `HorizontalRuler(margin = 0)`

File: REPLesent.scala [hunk 2]
```diff
@@ -92,6 +92,29 @@ case class REPLesent(
...
+        val s = line.toString
+        val t = s * (width / s.length)
+        blank * left + t + s.take(width - t.length) + blank * right
```
Comment (marconilanna) [inline REPLesent.scala]: You could just call `Alignment.fill` to literally fill in the blanks for you here :smile: Then you can make `blank` private again

File: REPLesent.scala [hunk 3]
```diff
@@ -92,6 +92,29 @@ case class REPLesent(
...
+      }
+    }
+
+    object HorizontalRule extends Repeat {
+      def apply(line: Line, margin: Int): String = {
+        val leftRight = margin/2
```
Comment (marconilanna) [inline REPLesent.scala]: What happens when `margin` is an odd number?

File: REPLesent.scala [hunk 4]
```diff
@@ -92,6 +92,29 @@ case class REPLesent(
       def apply(line: Line, margin: Int): String
     }
 
+    protected sealed trait Repeat extends Alignment {
+      protected def fill(line: Line, width: Int, left: Int, right: Int): String = {
+        val s = line.toString
+        val t = s * (width / s.length)
```
Comment (marconilanna) [inline REPLesent.scala]: No need for `val s` here: `val t = line * (...`

Comment (marconilanna) [inline REPLesent.scala]: You should always use `Line#length`, never `.toString.length`: the string representation of a `Line` may contain non-printable ANSI color escape codes that take up bytes on the string but do not occupy space on the screen.

File: REPLesent.scala (hunk 2 above)
Comment (marconilanna) [inline REPLesent.scala]: `.take(width - t.length)`: What is a challenge here is that the ruler pattern may contain ANSI color sequences. When you do that, you may be cutting an escape code in half, with unpredictable results. Which also means that you cannot rely on `t.length` to give you any meaningful information.

File: REPLesent.scala [hunk 5]
```diff
@@ -147,6 +170,8 @@ case class REPLesent(
       case s if s startsWith "| " => (s.drop(2), Centered)
       case s if s startsWith "> " => (s.drop(2), RightAligned)
       case s if s startsWith ">> " => (s.drop(3), RightFlushed)
+      case s if s startsWith "//" => (s.drop(2), FullScreenHorizontalRule)
+      case s if s startsWith "/" => (s.drop(1), HorizontalRule)
```
Comment (marconilanna) [inline REPLesent.scala]: Maybe for consistency, we should require that `//` and `/` be always followed by a space

File: test/REPLesentSpec.scala [hunk 6]
```diff
@@ -1354,6 +1354,89 @@ class REPLesentSpec extends FreeSpec {
...
+
+      val end = capture(replesent.next)
+      assert(end.output.isEmpty)
+      assert(end.error.nonEmpty)
+    }
+  }
+
+  "Horizontal Line:" - {
+    "Full Screen" in {
+      val (w, h) = (16, 3)
```
Comment (marconilanna) [inline test/REPLesentSpec.scala]: You should always test with odd and even widths.

File: test/resources/test_horizontal_rule_full_screen.txt [hunk 7]
```diff
@@ -0,0 +1,5 @@
+//-
```
Comment (marconilanna) [inline test/resources/test_horizontal_rule_full_screen.txt]: Yeah, I think a mandatory space separating the slash from the pattern would be aesthetically more pleasing

File: REPLesent.scala (hunk 4 above)
Comment (marconilanna) [inline REPLesent.scala]: If the line is empty, we gonna have a division by zero here (see, for instance, how `Config#fill` handles it). You should probably provide a reasonable default (`-` maybe?) if a pattern is not specified.

File: REPLesent.scala (hunk 2 above)
Comment (marconilanna) [inline REPLesent.scala]: You may have to use `takeWhile` and count chars ignoring escape codes

**Comment by daviscabral** [comment]
@marconilanna I couldn't fix the issue because I couldn't understand how to use `takeWhile` in that context (identifying the ANSI codes). Can you take a look and show how to do it? Thanks.

File: REPLesent.scala [hunk 8]
```diff
@@ -92,6 +92,23 @@ case class REPLesent(
       def apply(line: Line, margin: Int): String
     }
 
+    object HorizontalRuler extends Alignment {
+      def apply(line: Line, margin: Int): String = {
+        val width = space - margin
+        val content: String = line.toString * (width / line.length)
+        val newLine: Line = Line(content + content.take(width - content.length))
```
Comment (marconilanna) [inline REPLesent.scala]: Remember that the string `content` contains escape codes, so `content.length` can never be used to count the number of characters in `content`.

Comment (daviscabral) [inline REPLesent.scala]: Oops, good catch! :-)

File: REPLesent.txt [hunk 9]
```diff
@@ -1,3 +1,14 @@
...
+
+// \r___|
+// \r_|__
+// \r___|
+// _-+-
+// -=-=-=
+// \r-=\b-=-
+/ -+-
+/ \r-=-\b=-
+---
```
Comment (marconilanna) [inline REPLesent.txt]: This probably should not be here

File: REPLesent.txt [hunk 10]
```diff
@@ -123,6 +134,15 @@ Use \\_ for \_underline\s and \\! for \!reverse\s.
...
 ---
+| \gHorizontal Ruler
+
+// _-+-
+// -=-=-=
+// \r-=\b-=-
+/ -+-
+/ \r-=-\b=-
+
+---
```
Comment (marconilanna) [inline REPLesent.txt]: The intent of this file is to act as a tutorial. It would be better if you could add some description about how the feature works and how to use it. Use, for instance, the slide that explains line alignment as a reference.

File: test/REPLesentSpec.scala [hunk 11]
```diff
@@ -1354,6 +1354,94 @@ class REPLesentSpec extends FreeSpec {
       val slide2 = capture(replesent.next)
       assert(slide2.output === expected2)
       assert(slide2.error.isEmpty)
+
+      val end = capture(replesent.next)
+      assert(end.output.isEmpty)
+      assert(end.error.nonEmpty)
```
Comment (marconilanna) [inline test/REPLesentSpec.scala]: Why were these assertions added to the previous test? I don't think they are necessary here.

**Comment by marconilanna** [comment]
@daviscabral What I would like to see now are more extensive unit tests, with comprehensive edge and corner case and off-by-one coverage. You should, at the very least, test all eight combinations of the following:

* Even and odd screen sizes
* Even and odd slide sizes
* Even and odd ruler sizes

Ruler sizes of 2 and 3 with slide/screen sizes of 7 and 8 should give you some good numbers to work with.

Also missing are the following cases:

* Empty ruler patterns: just a `/ ` or `// ` by their own on a line
* Patterns that contain color codes, but do not start with one, like `/ -\g=`

Great work, thank you for the gentle contributions.
